source/SH3/graphics/msbmp.cpp

```diff
@@ -0,0 +1,100 @@
+/** @file
+ *
+ *  @copyright 2016  Palm Studios
+ *
+ *  @date 6-2-17
+ *
+ *  @author Jesse Buhagiar
+ */
+
+#include "SH3/graphics/msbmp.hpp"
+#include "SH3/system/log.hpp"
+
+#include <fstream>
+
+using namespace sh3_graphics;
+
+msbmp::load_result msbmp::Load(const std::string& path)
+{
+    std::ifstream   file;
+    bmp_file_header fheader; // File information header
+    bmp_info_header iheader;
+    load_result res;
+
+    res.set_error(result::BAD_LOAD);
+    file.open(path, std::ios::binary);
+
+    if(!file.is_open())
+    {
+        // If this happens we're in a bit of trouble, because we can't load the default texture..... We could crash here.
+        Log(LogLevel::ERROR, "msbmp::Load( ): Unable to open a handle to %s!", path.c_str());
+        return res;
+    }
+
+    file.read((char*)&fheader, sizeof(fheader)); // Read in the file header
+    if(fheader.tag[0] != 'B' || fheader.tag[1] != 'M')
+    {
+        Log(LogLevel::WARN, "msmbmp::Load( ): Warning: This file does not appear to be a BMP! Reason: BAD_HEADER!");
+        return res;
+    }
+
+    file.read((char*)&iheader, sizeof(iheader)); // Read in the information header
+    width = iheader.width;
+    height = iheader.height;
+    bpp = iheader.bpp;
+
+    data.resize(width * height * 3); // Some programs misformat this, so calculate it ourselves. This is usually 24-bit even for 8-bit paletted images!
+
+    // We can now differentiate if our bitmap is 8bpp or 24bpp
+    if(bpp == 8) // Paletted images are for fuckwits, I'm looking at you, Konami....
+    {
+        palette.resize(iheader.palette_size);
+        file.seekg(sizeof(fheader) + sizeof(iheader), std::ios_base::beg); // Seek to the palette
+        file.read((char*)&palette[0], iheader.palette_size);
+
+        // Slightly expensive, but we now copy each pixel into the data buffer by getting it's index in the palette.
+        // Seeing as this is done once, I doubt it matters too much, especially if we do this during load time
+        // and keep the image alive for the whole of the scene duration (there are 3 bmps total in SH3, not including
+        // our default 'error' texture)
+        for(int i = 0; i < width * height; i += 3)
+        {
+            std::uint32_t   colour; // 4-byte colour, stored in BGRA (where A is unused)
```

Comment:
Hahaha classic. I can only imagine what the original source for SH3 looks like. If you open up the executable in a debugger and search for strings, you can see some realllllly Japanese style English for the error messages that makes it impossible to interpret.